cql3/functions/functions.cc

```diff
@@ -684,6 +684,20 @@ function_call::raw::to_string() const {
     return format("{}({})", _name, join(", ", _terms));
 }
 
+namespace aggregate_fcts {
+
+template <typename Type>
+void impl_sum_function_for<Type>::validate_sum() {
+    Type short_sum = static_cast<Type>(_sum);
+    if (static_cast<accumulator_type>(short_sum) != _sum) {
+        log.warn(format("SUM overflow: 0x{:08x}_{:08x}_{:08x}_{:08x} -> {}",
```

Comment:
My main question is: Do we need this? Does cassandra produce the same warning?

If in CQL signed integer arithmetic is defined to be modular, why not just use std::make_unsigned to select the type and skip the warning?

Reply:
Regarding AVG - this agrees with what Cassandra produces (at least for the `int` type, not sure about the `bigint` type, but I'd consider it a bug if the internal accumulator overflows since the average of data of type T must always fit in T)

I can't say that SUM is "defined to be modular", but it behaves this way in Cassandra.
From the user's perspective I think it's bad to lose information about overflow.
Id we suppress any warning then I'd be happy if we add a way in CQL to declare a type type for the SUM that is different than the column type, e.g. `SELECT SUM(val) FROM cf AS bigint;`

Reply:
@avi / @slivne / @tzach / @nyh  what's you take on the above?

Reply:
Summoning also @GalitElad

Reply:
Better to throw an exception (and fail the query). There's no point in returning a modulo arithmetic sum.

The user can widen the calculation by using `SELECT SUM(val AS bigint) FROM tab`.

What do other databases do? Perhaps it makes sense for SUM to auto-widen, at least up to bigint.

Reply:
I'm all for either auto-widen the result (easy), or throw an exception if it doesn't fit (not very hard either)

FWIW, Cassandra decided on silently truncating the result.
See https://issues.apache.org/jira/browse/CASSANDRA-9674

Reply:
Auto-widen is not so easy, because it changes the prepared statement metadata. So we have to introduce a cluster feature, and hash the feature into prepared statement ID so we don't mix an old prepared statement with a new one.

Reply:
Though I'm not sure how would the clients handle possible overflow in the result.

I suggest that as a first step we throw an error since, as @avikivity mentioned, the user/app can promote the input to safely retrieve a larger result.

Then, as a follow-up, consider auto-promotion of the result type if feasible.

Reply:
But let's collect information about what other databases do.